# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/CanonicalBrowsingContext.h
# Commit: 30d2bf04719b
# Full Hash: 30d2bf04719bfb60095aaa7d679363e57d25ad92
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-06-13 03:21:52
# Regressor Bug: 1640019
# File Overlap Count: 1
# Description:
#   Bug 1640019 - Part 1: Support toplevel process switches outside of tabbrowser, r=mattwoodrow
#   
#   This new process switching behavior is only enabled for some browser elements,
#   which have specified a specific attribute. Turning this on for all browsers with
#   a `remote` attribute causes breakage in reftests.
# ==============================================================================

diff -r 3fefa1d023a6 -r 30d2bf04719b docshell/base/CanonicalBrowsingContext.h
--- a/docshell/base/CanonicalBrowsingContext.h	Fri Jun 12 17:02:45 2020 +0000
+++ b/docshell/base/CanonicalBrowsingContext.h	Fri Jun 12 16:52:00 2020 +0000
@@ -131,9 +131,15 @@
   void LoadURI(const nsAString& aURI, const LoadURIOptions& aOptions,
                ErrorResult& aError);
 
+  // Internal method to change which process a BrowsingContext is being loaded
+  // in. The returned promise will resolve when the process switch is completed.
+  //
+  // A VoidString() aRemoteType argument will perform a process switch into the
+  // parent process, and the method will resolve with a null BrowserParent.
   using RemotenessPromise = MozPromise<RefPtr<BrowserParent>, nsresult, false>;
-  RefPtr<RemotenessPromise> ChangeFrameRemoteness(const nsAString& aRemoteType,
-                                                  uint64_t aPendingSwitchId);
+  RefPtr<RemotenessPromise> ChangeRemoteness(const nsAString& aRemoteType,
+                                             uint64_t aPendingSwitchId,
+                                             bool aReplaceBrowsingContext);
 
   // Return a media controller from the top-level browsing context that can
   // control all media belonging to this browsing context tree. Return nullptr
@@ -175,13 +181,11 @@
 
     PendingRemotenessChange(CanonicalBrowsingContext* aTarget,
                             RemotenessPromise::Private* aPromise,
-                            uint64_t aPendingSwitchId)
-        : mTarget(aTarget),
-          mPromise(aPromise),
-          mPendingSwitchId(aPendingSwitchId) {}
+                            uint64_t aPendingSwitchId,
+                            bool aReplaceBrowsingContext);
 
     void Cancel(nsresult aRv);
-    void Complete(ContentParent* aContentParent);
+    void ProcessReady(ContentParent* aContentParent);
 
    private:
     ~PendingRemotenessChange();
@@ -191,6 +195,7 @@
     RefPtr<RemotenessPromise::Private> mPromise;
 
     uint64_t mPendingSwitchId;
+    bool mReplaceBrowsingContext;
   };
 
   friend class net::DocumentLoadListener;